packages/runtime/datastore/src/channelDeltaConnection.ts

```diff
@@ -75,25 +97,41 @@ export class ChannelDeltaConnection implements IDeltaConnection {
 		this.handler.setConnectionState(connected);
 	}
 
-	public process(
-		message: ISequencedDocumentMessage,
+	public processMessages(
+		message: ISequencedRuntimeMessageCore,
+		messageContents: IRuntimeMessageContents[],
 		local: boolean,
-		localOpMetadata: unknown,
-	) {
-		try {
-			// catches as data processing error whether or not they come from async pending queues
-			processWithStashedOpMetadataHandling(
-				message.contents,
-				localOpMetadata,
-				(contents, metadata) =>
-					this.handler.process({ ...message, contents }, local, metadata),
-			);
-		} catch (error) {
-			throw DataProcessingError.wrapIfUnrecognized(
-				error,
-				"channelDeltaConnectionFailedToProcessMessage",
-				message,
-			);
+	): void {
+		const newMessageContents = getMessagesWithStashedOpHandling(messageContents);
+		if (this.handler.processMessages !== undefined) {
+			try {
+				// catches as data processing error whether or not they come from async pending queues
+				this.handler.processMessages(message, newMessageContents, local);
+			} catch (error) {
+				throw DataProcessingError.wrapIfUnrecognized(
+					error,
+					"channelDeltaConnectionFailedToProcessMessages",
+				);
+			}
+			return;
+		}
+
+		for (const { contents, localOpMetadata, clientSequenceNumber } of newMessageContents) {
+			const compatMessage: ISequencedDocumentMessage = {
+				...message,
+				contents,
+				clientSequenceNumber,
+			};
+			try {
+				// catches as data processing error whether or not they come from async pending queues
+				this.handler.process(compatMessage, local, localOpMetadata);
+			} catch (error) {
+				throw DataProcessingError.wrapIfUnrecognized(
```

Comment:
Feels like it's a bit easier to follow logic if you have one giant try/catch around whole function instead of repeating it.

Reply:
Changed it.